File: src/verifier.ts
import type { GitHubClient, IssueRef, VerificationResult } from "./types.js";
import type { PatternMatcher } from "./pattern-matcher.js";

const FIX_REFERENCE_PATTERN = /\b(?:fixed|resolved|closed|addressed)\s+(?:in|by|via)?\s*(?:pr\s*)?#(\d+)/i;

/**
 * Extracts the PR number a piece of text claims fixed an issue, e.g.
 * "Fixed in PR #52", "fixed by #52", "resolved in #52". Returns the last
 * match found (closing comments are usually the most authoritative and
 * tend to come after earlier discussion in the same body/thread).
 */
export function extractFixReference(text: string): number | null {
  const matches = [...text.matchAll(new RegExp(FIX_REFERENCE_PATTERN, "gi"))];
  if (matches.length === 0) return null;
  const last = matches[matches.length - 1];
  const raw = last?.[1];
  if (!raw) return null;
  const num = Number.parseInt(raw, 10);
  return Number.isFinite(num) ? num : null;
}

export interface PatternSpec {
  path: string;
  pattern: string;
}

/**
 * Verifies whether issues closed as "fixed" actually have a merged fix.
 * The core, unambiguous check: tracker state (closed, cites a fix PR) vs.
 * actual PR merge state. An optional PatternMatcher adds a secondary,
 * best-effort check for whether the cited vulnerable code pattern is
 * still present at HEAD.
 */
export class TrackerVerifier {
  constructor(
    private readonly client: GitHubClient,
    private readonly patternMatcher?: PatternMatcher,
  ) {}

  async verify(ref: IssueRef, patternSpec?: PatternSpec): Promise<VerificationResult> {
    const issue = await this.client.getIssue(ref);
    const issueUrl = `https://github.com/${ref.owner}/${ref.repo}/issues/${ref.number}`;

    if (issue.state === "open") {
      return {
        issue: { number: issue.number, title: issue.title, state: issue.state, htmlUrl: issueUrl },
        citedPullRequest: null,
        patternCheck: null,
        verdict: "UNVERIFIED",
        reason: "Issue is still open; no fix has been claimed.",
      };
    }

    const comments = await this.client.getIssueComments(ref);
    const searchText = [issue.body ?? "", ...comments.map((c) => c.body)].join("\n");
    const prNumber = extractFixReference(searchText);

    if (prNumber === null) {
      return {
        issue: { number: issue.number, title: issue.title, state: issue.state, htmlUrl: issueUrl },
        citedPullRequest: null,
        patternCheck: null,
        verdict: "UNVERIFIED",
        reason: "Issue is closed but no fix PR reference (e.g. \"Fixed in #N\") was found in its body or comments.",
      };
    }

    const pr = await this.client.getPullRequest(ref.owner, ref.repo, prNumber);
    const citedPullRequest = {
      number: pr.number,
      state: pr.state,
      merged: pr.merged,
      htmlUrl: pr.htmlUrl,
    };

    if (!pr.merged) {
      return {
        issue: { number: issue.number, title: issue.title, state: issue.state, htmlUrl: issueUrl },
        citedPullRequest,
        patternCheck: null,
        verdict: "MISMATCH",
        reason: `Issue is closed and cites PR #${prNumber} as the fix, but that PR is ${pr.state} and was never merged.`,
      };
    }

    if (!patternSpec || !this.patternMatcher) {
      return {
        issue: { number: issue.number, title: issue.title, state: issue.state, htmlUrl: issueUrl },
        citedPullRequest,
        patternCheck: null,
        verdict: "MATCH",
        reason: `Issue is closed and PR #${prNumber} is merged.`,
      };
    }

    const patternCheck = await this.patternMatcher.check(ref.owner, ref.repo, patternSpec.path, patternSpec.pattern);

    if (patternCheck.found === true) {
      return {
        issue: { number: issue.number, title: issue.title, state: issue.state, htmlUrl: issueUrl },
        citedPullRequest,
        patternCheck,
        verdict: "MISMATCH",
        reason: `PR #${prNumber} is merged, but the cited vulnerable pattern is still present in ${patternSpec.path} at HEAD.`,
      };
    }

    if (patternCheck.found === null) {
      return {
        issue: { number: issue.number, title: issue.title, state: issue.state, htmlUrl: issueUrl },
        citedPullRequest,
        patternCheck,
        verdict: "MATCH",
        reason: `Issue is closed and PR #${prNumber} is merged. Pattern check inconclusive: ${patternCheck.note ?? "unknown"}.`,
      };
    }

    return {
      issue: { number: issue.number, title: issue.title, state: issue.state, htmlUrl: issueUrl },
      citedPullRequest,
      patternCheck,
      verdict: "MATCH",
      reason: `Issue is closed, PR #${prNumber} is merged, and the cited pattern is confirmed absent from ${patternSpec.path} at HEAD.`,
    };
  }
}

File: python/src/shimguard/verifier.py
"""
Ported from src/verifier.ts (TrackerVerifier, extractFixReference). Same
core, unambiguous check: tracker state (closed, cites a fix PR) vs. actual
PR merge state. Same optional secondary check via a PatternMatcher for
whether the cited vulnerable code pattern is still present at HEAD.
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import List, Optional

from .pattern_matcher import PatternMatcher
from .types import (
    GitHubClient,
    IssueRef,
    IssueSummary,
    PullRequestSummary,
    VerificationResult,
)

# Matched against issue body/comment text fetched live from the repo being
# audited, which may be adversarial. The connector text between the keyword
# and "#N" is bounded to 80 chars via a single non-backtracking character
# class instead of chained optional/greedy groups -- the same ReDoS-safe
# shape as src/verifier.ts's FIX_REFERENCE_PATTERN, ported verbatim.
FIX_REFERENCE_PATTERN = re.compile(
    r"\b(?:fixed|resolved|closed|addressed)\b[^#\n]{0,80}#(\d+)", re.IGNORECASE
)


def extract_fix_reference(text: str) -> Optional[int]:
    """
    Extracts the PR number a piece of text claims fixed an issue, e.g.
    "Fixed in PR #52", "fixed by #52", "resolved in #52". Returns the last
    match found (closing comments are usually the most authoritative and
    tend to come after earlier discussion in the same body/thread).
    """
    matches: List["re.Match[str]"] = list(FIX_REFERENCE_PATTERN.finditer(text))
    if not matches:
        return None
    raw = matches[-1].group(1)
    try:
        return int(raw)
    except ValueError:
        return None


@dataclass(frozen=True)
class PatternSpec:
    path: str
    pattern: str


class TrackerVerifier:
    def __init__(self, client: GitHubClient, pattern_matcher: Optional[PatternMatcher] = None) -> None:
        self.client = client
        self.pattern_matcher = pattern_matcher

    def verify(self, ref: IssueRef, pattern_spec: Optional[PatternSpec] = None) -> VerificationResult:
        issue = self.client.get_issue(ref.owner, ref.repo, ref.number)
        issue_url = f"https://github.com/{ref.owner}/{ref.repo}/issues/{ref.number}"
        issue_summary = IssueSummary(
            number=issue.number, title=issue.title, state=issue.state, html_url=issue_url
        )

        if issue.state == "open":
            return VerificationResult(
                issue=issue_summary,
                cited_pull_request=None,
                pattern_check=None,
                verdict="UNVERIFIED",
                reason="Issue is still open; no fix has been claimed.",
            )

        comments = self.client.get_issue_comments(ref.owner, ref.repo, ref.number)
        search_text = "\n".join([issue.body or ""] + [c.body for c in comments])
        pr_number = extract_fix_reference(search_text)

        if pr_number is None:
            return VerificationResult(
                issue=issue_summary,
                cited_pull_request=None,
                pattern_check=None,
                verdict="UNVERIFIED",
                reason=(
                    'Issue is closed but no fix PR reference (e.g. "Fixed in #N") '
                    "was found in its body or comments."
                ),
            )

        pr = self.client.get_pull_request(ref.owner, ref.repo, pr_number)
        cited_pull_request = PullRequestSummary(
            number=pr.number, state=pr.state, merged=pr.merged, html_url=pr.html_url
        )

        if not pr.merged:
            return VerificationResult(
                issue=issue_summary,
                cited_pull_request=cited_pull_request,
                pattern_check=None,
                verdict="MISMATCH",
                reason=(
                    f"Issue is closed and cites PR #{pr_number} as the fix, but that PR "
                    f"is {pr.state} and was never merged."
                ),
            )

        if pattern_spec is None or self.pattern_matcher is None:
            return VerificationResult(
                issue=issue_summary,
                cited_pull_request=cited_pull_request,
                pattern_check=None,
                verdict="MATCH",
                reason=f"Issue is closed and PR #{pr_number} is merged.",
            )

        pattern_check = self.pattern_matcher.check(
            ref.owner, ref.repo, pattern_spec.path, pattern_spec.pattern
        )

        if pattern_check.found is True:
            return VerificationResult(
                issue=issue_summary,
                cited_pull_request=cited_pull_request,
                pattern_check=pattern_check,
                verdict="MISMATCH",
                reason=(
                    f"PR #{pr_number} is merged, but the cited vulnerable pattern is "
                    f"still present in {pattern_spec.path} at HEAD."
                ),
            )

        if pattern_check.found is None:
            return VerificationResult(
                issue=issue_summary,
                cited_pull_request=cited_pull_request,
                pattern_check=pattern_check,
                verdict="MATCH",
                reason=(
                    f"Issue is closed and PR #{pr_number} is merged. Pattern check "
                    f"inconclusive: {pattern_check.note or 'unknown'}."
                ),
            )

        return VerificationResult(
            issue=issue_summary,
            cited_pull_request=cited_pull_request,
            pattern_check=pattern_check,
            verdict="MATCH",
            reason=(
                f"Issue is closed, PR #{pr_number} is merged, and the cited pattern is "
                f"confirmed absent from {pattern_spec.path} at HEAD."
            ),
        )
